clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -0,0 +1,331 @@
+            gen_exit(3),
+            gen_instr(1, 1),
+            gen_branch(1, 2),
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 3),
+            gen_marker(2, TRACE_MARKER_TYPE_TIMESTAMP, 0),
+            gen_instr(2, 4),
+        };
+        for (const auto &memref : memrefs) {
+            checker.process_memref(memref);
+        }
+        for (const auto &error : checker.errors) {
+            std::cerr << "Unexpected error: " << error << "\n";
+        }
+        if (!checker.errors.empty())
+            return false;
+    }
+    return true;
+}
+
+bool
+check_sane_control_flow()
+{
+    // Negative simple test.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1),
+            gen_instr(1, 3),
+        };
+        for (const auto &memref : memrefs) {
+            checker.process_memref(memref);
+        }
+        if (checker.errors.size() != 1 ||
+            checker.errors[0] != "Non-explicit control flow has no marker") {
+            std::cerr << "Failed to catch bad control flow\n";
+            return false;
+        }
+    }
+    // String loop.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_instr_type(TRACE_TYPE_INSTR_NO_FETCH, 1, 1),
+            gen_instr_type(TRACE_TYPE_INSTR_NO_FETCH, 1, 1),
+            gen_instr_type(TRACE_TYPE_INSTR_NO_FETCH, 1, 1),
+            gen_instr_type(TRACE_TYPE_INSTR_NO_FETCH, 1, 1),
+            gen_instr(1, 2),
+        };
+        for (const auto &memref : memrefs) {
+            checker.process_memref(memref);
+        }
+        for (const auto &error : checker.errors) {
+            std::cerr << "Unexpected error: " << error << "\n";
+        }
+        if (!checker.errors.empty())
+            return false;
+    }
+    // Kernel-mediated.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1),
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
+            gen_instr(1, 101),
+        };
+        for (const auto &memref : memrefs) {
+            checker.process_memref(memref);
+        }
+        for (const auto &error : checker.errors) {
+            std::cerr << "Unexpected error: " << error << "\n";
+        }
+        if (!checker.errors.empty())
+            return false;
+    }
+    return true;
+}
+
+bool
+check_kernel_xfer()
+{
+#ifdef UNIX
+    // Return to recorded interruption point.
+    {
+        checker_no_abort_t checker(true /*offline*/);
+        std::vector<memref_t> memrefs = {
+            gen_instr(1, 1),
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_EVENT, 2),
+            gen_instr(1, 101),
+            // XXX: This marker value is actually not guaranteed, yet the checker
+            // requires it and the view tool prints it.
+            gen_marker(1, TRACE_MARKER_TYPE_KERNEL_XFER, 102),
```

Comment:
Will another test case for nested signals be useful?

Reply:
Yes, the plan is to add a nested signals check: but right now that check does not exist in invariant_checker (partly b/c it is really hard in the presence of longjmp).  That will be a future PR.